fs-storage/src/file_storage.rs

```diff
@@ -176,12 +152,97 @@ where
+    fn set(&mut self, key: K, value: V) {
+        self.data.entries.insert(key, value);
+        self.modified = std::time::SystemTime::now();
+    }
+
+    /// Remove an entry from the internal mapping given a key
+    fn remove(&mut self, id: &K) -> Result<()> {
+        self.data.entries.remove(id).ok_or_else(|| {
+            ArklibError::Storage(self.label.clone(), "Key not found".to_owned())
+        })?;
+        self.modified = std::time::SystemTime::now();
+        Ok(())
+    }
+
+    /// Compare the timestamp of the storage file
+    /// with the timestamp of the in-memory storage and the last written
+    /// to time to determine if either of the two requires syncing.
+    fn sync_status(&self) -> Result<SyncStatus> {
+        let file_updated = fs::metadata(&self.path)?.modified()?;
+
+        // Determine the synchronization status based on the modification times
+        // Conditions:
+        // 1. If both the in-memory storage and the storage on disk have been modified
+        //    since the last write, then the storage is diverged.
+        // 2. If only the in-memory storage has been modified since the last write,
+        //    then the storage on disk is stale.
+        // 3. If only the storage on disk has been modified since the last write,
+        //    then the in-memory storage is stale.
+        // 4. If neither the in-memory storage nor the storage on disk has been modified
+        //    since the last write, then the storage is in sync.
+        let status = match (
+            self.modified > self.written_to_disk,
+            file_updated > self.written_to_disk,
+        ) {
+            (true, true) => SyncStatus::Diverge,
+            (true, false) => SyncStatus::StorageStale,
+            (false, true) => SyncStatus::MappingStale,
+            (false, false) => SyncStatus::InSync,
+        };
+
+        log::info!("{} sync status is {}", self.label, status);
+        Ok(status)
+    }
+
+    /// Sync the in-memory storage with the storage on disk
+    fn sync(&mut self) -> Result<()> {
+        match self.sync_status()? {
+            SyncStatus::InSync => Ok(()),
+            SyncStatus::MappingStale => self.read_fs().map(|_| ()),
+            SyncStatus::StorageStale => self.write_fs().map(|_| ()),
+            SyncStatus::Diverge => {
+                let data = self.load_fs_data()?;
+                self.merge_from(&data)?;
+                self.write_fs()?;
+                Ok(())
+            }
+        }
+    }
+
+    /// Read the data from file
+    fn read_fs(&mut self) -> Result<&BTreeMap<K, V>> {
+        let data = self.load_fs_data()?;
+
+        // Update file storage with loaded data
         self.modified = fs::metadata(&self.path)?.modified()?;
+        self.written_to_disk = self.modified;
+        self.data = data;
 
-        Ok(data.entries)
+        Ok(&self.data.entries)
     }
 
-    /// Write the data to the storage file
+    /// Write the data to file
```

Comment:
We should also mention the trick of manually modifying the timestamp and why we need it as a note in the doc comment of this method